feat(webchat/guide): add render dropdown menu guide (#438)
botpress/docs: webchat/interact/dropdown-menus.mdx

---

> title: Dropdown menus

Reviewer: I would consider changing the title to the imperative - something like "Send options as a dropdown menu"

you might also want to put this under **Studio** > **Guides** > **How-to**. I understand the thinking behind putting it under Webchat, but nothing in this guide directly touches Webchat

Author: done

> description: Use Capture Information for automatic dropdowns, or an Execute Code card to show a dropdown with any number of options, including fewer than six.

Reviewer: this is a bit long/wordy for a description

Author: tbh ill delete it cuz Studio guides dont have one, ig consistency

---

> You can present options as a **dropdown** instead of buttons. This guide covers the built-in behaviour with [Capture Information](/studio/concepts/cards/capture-information) and an approach **using code** when you need a dropdown with **fewer than six** options.

Reviewer: ```suggestion
You can present options as a dropdown instead of buttons. This guide covers the built-in behaviour with [Capture Information](/studio/concepts/cards/capture-information) and an approach using code when you need a dropdown with fewer than six options.
```

the docs are not fully consistent with bolding the main topic in the introductory paragraph, but none of the how-to guides in the Webchat and Studio sections do it, so let's not do it here

<Info>
    You will need:

    1. A [Workflow](/studio/concepts/workflows) in [Botpress Studio](/studio/introduction)

>     2. Familiarity with the [Execute Code](/studio/concepts/cards/execute-code) card for the manual method

Reviewer: ```suggestion
    - A [published bot](/get-started/quick-start)
    - Familiarity with JavaScript (if you use [method 2](#method-2-manual-dropdown-execute-code)
```

This is the simplest approach when you have enough options to trigger the dropdown UI.

## Method 2: Manual dropdown (Execute Code)

To show a **dropdown even when you have five or fewer** options, or to fully control the payload, send a message with `type: 'dropdown'` from an [Execute Code](/studio/concepts/cards/execute-code) card.

### Send the dropdown message

```javascript
await client.createMessage({
  userId: event.botId,
  conversationId: event.conversationId,
  tags: {},
  type: 'dropdown',
  payload: {
    text: 'Select an item:', // Shown above the control
    options: [
      { label: 'Dropdown 1', value: 'dd1' },
      { label: 'Dropdown 2', value: 'dd2' },
    ],
  },
})
```

1. **`payload.text`**: Label or instruction shown with the control.
2. **`payload.options`**: Each entry is `{ label: string, value: string }`.
    1. **`label`**: Text the user sees (for example a product name).
    2. **`value`**: Value stored or matched in logic (for example an SKU). They can match or differ.

<Tip>
    In Execute Code, you can read context from the [event object](/studio/guides/advanced/event-properties).
</Tip>

### Complete the Workflow after the dropdown

> After the dropdown is sent, the user must select an option. Typical pieces:

Reviewer: ```suggestion
After the dropdown is sent, the user must select an option. Here are some typical ways to handle this:
```

1. **Wait for input.** Add a **Wait for User Input** step so the Workflow pauses until the user interacts. See [Wait for User Input](/studio/concepts/cards/capture-information#wait-for-user-input) on the Capture Information page (or combine with your next card’s capture behaviour as you prefer).

2. **If the user sends free text instead of using the dropdown**, use an [Expression](/studio/concepts/cards/flow-logic#expression) (or similar) **transition** so that:
    1. If there is **no** selection payload, **loop back** to the same node (or show a short reminder and stay on the prompt).

>     2. Example condition for “no dropdown value”: `!event.payload.value`

Reviewer: ```suggestion
2. If the user sends free text instead of using the dropdown, use an [Expression Card](/studio/concepts/cards/flow-logic#expression) (or similar) so you can loop back to the same Node and enforce that they pick an option.
Here's an example condition you can add to the Expression Card to evaluate that their response didn't contain a choice from the dropdown: `!event.payload.value`
```

3. **Store the selection.** In an **Execute Code** card after a valid selection:

    ```javascript
    workflow.varName = event.payload.value
    ```

>     [Create a Workflow variable](/studio/concepts/variables/scopes/workflow) (for example `varName`, type **String**) and grant access in code as needed ([variables in code](/studio/concepts/variables/in-code)).

Reviewer: ```suggestion
3. Use an **Execute Code** Card to store their choice. For example:

    ```javascript
    workflow.varName = event.payload.value
    ```

    Remember to [create a Workflow variable](/studio/concepts/variables/scopes/workflow) (for example `varName`, type **String**) and grant access in code as needed ([variables in code](/studio/concepts/variables/in-code)).
```

> Wire these cards in order in the same or connected nodes so the bot always waits, validates, then saves before continuing.

Reviewer: ```suggestion
Wire these cards in order in the same Node (or connected Nodes) so the bot always waits, validates, then saves before continuing.
```

## Customize the placeholder text (CSS)

Classes for styling dropdowns are listed under **Use custom styles** on the [appearance settings](/webchat/get-started/configure-your-webchat#use-custom-styles) page.

To replace the default **Select…** style label with your own, add custom CSS under **Bot Appearance** → **Styles** in the Studio, for example:

```css
/* Hide the original "Select..." text */
.bpMessageBlocksDropdownButtonText {
  font-size: 0;
  color: transparent;
}

.bpMessageBlocksDropdownButtonText::before {
  content: "Choose an option...";
  font-size: 0.8rem;
  color: #7e7c7c;
}
```

Adjust `content`, `font-size`, and `color` to match your brand.

<Check>

>     You can rely on **Capture Information** for dropdowns when you have **more than five** choices, or use **`createMessage`** with **`type: 'dropdown'`** in **Execute Code** whenever you need a dropdown with **fewer options** or custom `label` and `value` pairs.

Reviewer: ```suggestion
    You can rely on **Capture Information** for dropdowns when you have more than five choices, or use `createMessage` with `type: 'dropdown'` in an Execute Code Card whenever you need a dropdown with fewer options or custom `label` and `value` pairs.
```
